File: DesignPatterns/Structural/Flyweight/CounterStrike.java
/*
Diffrence between factory pattern is that it stores the created object and reuses it. Fac
*/
import java.util.Random; 
import java.util.HashMap;


interface Player {
  public void assignWeapon(String weapon);
  public void mission();
}

class Terrorist implements Player {

  private final String TASK;
  private String weapon;

  public Terrorist(){
    TASK = "Plant a Bomb";
  }

  public void assignWeapon(String weapon) {
    this.weapon = weapon;
  }

  public void mission(){
    System.out.println("Terrorist with weapon " + weapon + " | " + "task is " + TASK);
  }
}


class CounterTerrorist implements Player{

  private final String TASK;
  private String weapon;

  public CounterTerrorist(){
    TASK = "Diffuse the bomb";
  }

  public void assignWeapon(String weapon) {
    this.weapon = weapon;
  }

  public void mission(){
    System.out.println("CounterTerrorist with weapon " + weapon + " | " + "task is " + TASK);
  }
}


class PlayerFactory {
  private static HashMap<String, Player> hm = new HashMap<String, Player>();

  public static Player getPlayer(String type){
    Player p = null;

    if(hm.containsKey(type)){
      p = hm.get(type);
    } else {
      switch(type){
        case "Terrorist": 
          System.out.println("Terrorist Created");
          p = new Terrorist();
          break;
        case "CounterTerrorist":
          System.out.println("CounterTerrorist Created");
          p = new CounterTerrorist();
          break;
        default:
          System.out.println("Invalid Type");
      }
      hm.put(type, p);
    }
    return p;
  }
}


public class CounterStrike {
  private static String[] playerType = {"Terrorist", "CounterTerrorist"};
  private static String[] weapons = {"AK-47", "Maverick", "Gut Knife", "Desert Eagle"};
  public static void main(String[] args) {
    for(int i=0; i<10; i++) {

      Player p = PlayerFactory.getPlayer(getRandomPlayerType());
      p.assignWeapon(getRandomWeapon());
      p.mission();
    }
  }

  private static String getRandomPlayerType(){
    Random r = new Random();
    int randInt = r.nextInt(playerType.length);
    return playerType[randInt];
  }

  private static String getRandomWeapon(){
    Random r = new Random();
    int randInt = r.nextInt(weapons.length);
    return weapons[randInt];
  }


}

File: DesignPatterns/Structural/Flyweight/Ruby/counter_strike.rb
class Player
  attr_reader :task, :weapon

  def initialize(task)
    @task = task
  end

  def assign_weapon(weapon)
    raise NotImplementedError
  end

  def mission
    raise NotImplementedError
  end
end

class Terrorist < Player
  def initialize
    super('Plant a Bomb')
  end

  def assign_weapon(weapon)
    @weapon = weapon
  end

  def mission
    puts "Terrorist with weapon #{@weapon} | task is #{@task}"
  end
end

class CounterTerrorist < Player
  def initialize
    super('Diffuse the bomb')
  end
  def assign_weapon(weapon)
    @weapon = weapon
  end

  def mission
    puts "CounterTerrorist with weapon #{@weapon} | task is #{@task}"
  end
end

module Constants
  TERRORIST = "TERRORIST".freeze
  COUNTER_TERRORIST = "COUNTER_TERRORIST".freeze
end

class PlayerFactory
  @@pool = {}

  def self.fetch_player(type)
    if @@pool.has_key?(type)
      @@pool[type]
    else
      case type
      when Constants::TERRORIST then @@pool[type] = Terrorist.new
      when Constants::COUNTER_TERRORIST then @@pool[type] = CounterTerrorist.new
      end
    end
  end
end

p1 = PlayerFactory.fetch_player(Constants::TERRORIST)
puts p1
p1.assign_weapon('AK-47')
p1.mission

p2 = PlayerFactory.fetch_player(Constants::COUNTER_TERRORIST)
puts p2
p2.assign_weapon('Maverick')
p2.mission

p3 = PlayerFactory.fetch_player(Constants::TERRORIST)
puts p3
p3.assign_weapon('Gut Knife')
p3.mission